CU-86a1gfwx5 - NEON3 - Implement Settings Screen Basic Structure

ricardoak25 opened the pull request.

ricardoak25 committed:
86a1gfwx5 - NEON3 - Implement Settings Screen Basic Structure

lllwvlvwlll commented:
Task linked: [CU-86a1gfwx5 NEON3 - Implement Settings Screen Basic Structure](https://app.clickup.com/t/86a1gfwx5)

raulduartep commented on src/renderer/src/components/Sidebar/SettingsSidebarLink.tsx:
Change the location of this file, it is only used by Settings, so place it in the same folder as the page

raulduartep commented on src/renderer/src/components/Sidebar/SettingsSidebarLink.tsx:
```diff
@@ -0,0 +1,32 @@
+import { cloneElement, useEffect, useState } from 'react'
+
+export const SettingsSidebarLink = ({ id, icon, title, onClick, selectedId }): JSX.Element => {
+  const [isActive, setIsActive] = useState<boolean>(id === selectedId)
```
It is better to receive a prop called `active` and who is rendering it should pass this prop, for example: 

```
<SettingsSidebarLink
    title={t('personalisationOption.networkConfiguration')}
    icon={<Tb3DCubeSphere />}
    onClick={() => handlePersonalisationClick(SettingsPersonalisationOption.NETWORK_CONFIGURATION)}
    active={selectedPersonalisationOption === SettingsPersonalisationOption.NETWORK_CONFIGURATION}
```

And you do that, you wont need the useState, useEffect

raulduartep commented on src/renderer/src/components/Sidebar/SettingsSidebarLink.tsx:
```diff
@@ -0,0 +1,32 @@
+  return (
+    <li className="my-3">
+      <button
+        onClick={() => handleClick()}
```
If you follow the instructions above you can change it to:
```suggestion
        onClick={onClick}
```

raulduartep commented on src/renderer/src/components/Sidebar/SettingsSidebarLink.tsx:
```diff
@@ -0,0 +1,32 @@
+import { cloneElement, useEffect, useState } from 'react'
+
+export const SettingsSidebarLink = ({ id, icon, title, onClick, selectedId }): JSX.Element => {
```
Create a type for the props
```suggestion
type TProps = {
icon: JSX.Element
title: string
onClick: MouseEventHandler<HTMLButtonElement>
}

export const SettingsSidebarLink = ({ id, icon, title, onClick, selectedId }): JSX.Element => {
...component
}
```

raulduartep commented on src/renderer/src/components/Sidebar/SettingsSidebarLink.tsx:
```diff
@@ -0,0 +1,32 @@
+    <li className="my-3">
+      <button
+        onClick={() => handleClick()}
+        className={`tab-button p-2 pl-3 w-full flex border-l-[3px] justify-content-end ${
```
Use rem instead of px

raulduartep commented on src/renderer/src/components/Sidebar/SettingsSidebarLink.tsx:
```diff
@@ -0,0 +1,32 @@
+    <li className="my-3">
+      <button
+        onClick={() => handleClick()}
+        className={`tab-button p-2 pl-3 w-full flex border-l-[3px] justify-content-end ${
+          isActive
+            ? 'border-l-neon bg-asphalt stroke-neon text-neon'
+            : 'border-l-transparent cursor-pointer opacity-60 hover:border-l-neon hover:bg-asphalt hover:opacity-100 hover:stroke-neon hover:text-neon'
+        }`}
```
You can use the StyleHelper.mergeStyles to create this condicional
```suggestion
        className={StyleHelper.mergeStyles("tab-button p-2 pl-3 w-full flex border-l-[3px] justify-content-end", {
        "border-l-neon bg-asphalt stroke-neon text-neon": active,
        "border-l-transparent cursor-pointer opacity-60 hover:border-l-neon hover:bg-asphalt hover:opacity-100 hover:stroke-neon hover:text-neon": !active
        })}
```

raulduartep commented on src/renderer/src/components/Sidebar/SettingsSidebarLink.tsx:
```diff
@@ -0,0 +1,32 @@
+          className: `w-5 h-5 object-contain`,
+        })}
+
+        <span className="ml-3 -mt-0.5 text-white">{title}</span>
```
Why you need -mt-0.5 ?

raulduartep commented on src/renderer/src/routes/pages/Settings/index.tsx:
```diff
@@ -0,0 +1,141 @@
+
+import { SettingsSidebarLink } from '../../../components/Sidebar/SettingsSidebarLink'
+
+enum SettingsSidebarOption {
+  PERSONALISATION = 1,
+  SECURITY = 2,
+}
+
+enum SettingsPersonalisationOption {
+  NETWORK_CONFIGURATION = 1,
+  CURRENCY = 2,
+  LANGUAGE = 3,
+  THEME = 4,
+  RELEASE_NOTES = 5,
+}
+
+enum SettingsSecurityOption {
+  ENCRYPT_KEY = 1,
+  RECOVER_WALLET = 2,
+  BACKUP_WALLET = 3,
+}
```
Add the prefix "E" to the name of the enumeration, because it is more descriptive when using that it is an enum, for example, ESettingsSidebarOption, etc.

Also, you can change the Numeric Enum for String Enum, i think it is more descriptive,  for example:

```
enum SettingsPersonalisationOption {
  NETWORK_CONFIGURATION="NETWORK_CONFIGURATION",
  CURRENCY="CURRENCY",
  LANGUAGE="LANGUAGE",
  THEME="THEME",
  RELEASE_NOTES="RELEASE_NOTES",
}
```

raulduartep commented on src/renderer/src/routes/pages/Settings/index.tsx:
```diff
@@ -0,0 +1,141 @@
+
+import { SettingsSidebarLink } from '../../../components/Sidebar/SettingsSidebarLink'
+
+enum SettingsSidebarOption {
```
Rename it to SettingsTabOption

raulduartep commented on src/renderer/src/routes/pages/Settings/index.tsx:
```diff
@@ -0,0 +1,141 @@
+          <nav className="flex flex-row justify-between h-15 w-full mb-5 text-[14px]">
+            <ul className="max-w-full w-full">
+              {selectedSidebarOption === SettingsSidebarOption.PERSONALISATION && (
+                <div>
+                  <SettingsSidebarLink
+                    id={SettingsPersonalisationOption.NETWORK_CONFIGURATION}
+                    title={t('personalisationOption.networkConfiguration')}
+                    icon={<Tb3DCubeSphere />}
+                    onClick={() => handlePersonalisationClick(SettingsPersonalisationOption.NETWORK_CONFIGURATION)}
+                    selectedId={selectedPersonalisationOption}
+                  />
+                  <SettingsSidebarLink
+                    id={SettingsPersonalisationOption.CURRENCY}
+                    title={t('personalisationOption.currency')}
+                    icon={<BsCash />}
+                    onClick={() => handlePersonalisationClick(SettingsPersonalisationOption.CURRENCY)}
+                    selectedId={selectedPersonalisationOption}
+                  />
+                  <SettingsSidebarLink
+                    id={SettingsPersonalisationOption.LANGUAGE}
+                    title={t('personalisationOption.language')}
+                    icon={<MdOutlineLanguage />}
+                    onClick={() => handlePersonalisationClick(SettingsPersonalisationOption.LANGUAGE)}
+                    selectedId={selectedPersonalisationOption}
+                  />
+                  <SettingsSidebarLink
+                    id={SettingsPersonalisationOption.THEME}
+                    title={t('personalisationOption.theme')}
+                    icon={<RiShapesLine />}
+                    onClick={() => handlePersonalisationClick(SettingsPersonalisationOption.THEME)}
+                    selectedId={selectedPersonalisationOption}
+                  />
+                  <SettingsSidebarLink
+                    id={SettingsPersonalisationOption.RELEASE_NOTES}
+                    title={t('personalisationOption.releaseNotes')}
+                    icon={<MdOutlineListAlt />}
+                    onClick={() => handlePersonalisationClick(SettingsPersonalisationOption.RELEASE_NOTES)}
+                    selectedId={selectedPersonalisationOption}
+                  />
+                </div>
```
As you have the enum, you can use a loop here so you don't have to create one by one.

Outside the component create a new const for describe the title and icon for each option:

```
const infosByOption: Record<SettingsPersonalisationOption | SettingsSecurityOption, {title: string, icon: JSX.Element}> = {
   ENCRYPT_KEY: {
    title: 'Encrypt Key',
    icon: <MdOutlineLock />,
   },
...others
}
```

Now you can use a loop to render each option
```
             {
             selectedSidebarOption === SettingsSidebarOption.PERSONALISATION && Object.keys().map(option => {
             const info = infosByPersonalisationOption[option as SettingsPersonalisationOption]

             return (<SettingsSidebarLink
                    key={option}
                    title={info.title}
                    icon={info.icon}
                    onClick={() => handlePersonalisationClick(option)}
                    active={selectedSecurityOption === option}
                  />)
             })
             }
```

raulduartep commented on src/renderer/src/routes/pages/Settings/index.tsx:
```diff
@@ -0,0 +1,141 @@
+                  />
+                </div>
+              )}
+              {selectedSidebarOption === SettingsSidebarOption.SECURITY && (
+                <div>
+                  <SettingsSidebarLink
+                    id={SettingsSecurityOption.ENCRYPT_KEY}
+                    title={t('securityOption.encryptKey')}
+                    icon={<MdOutlineLock />}
+                    onClick={() => handleSecurityClick(SettingsSecurityOption.ENCRYPT_KEY)}
+                    selectedId={selectedSecurityOption}
+                  />
+                  <SettingsSidebarLink
+                    id={SettingsSecurityOption.RECOVER_WALLET}
+                    title={t('securityOption.recoverWallet')}
+                    icon={<TbReload />}
+                    onClick={() => handleSecurityClick(SettingsSecurityOption.RECOVER_WALLET)}
+                    selectedId={selectedSecurityOption}
+                  />
+                  <SettingsSidebarLink
+                    id={SettingsSecurityOption.BACKUP_WALLET}
+                    title={t('securityOption.backupWallet')}
+                    icon={<MdOutlineSave />}
+                    onClick={() => handleSecurityClick(SettingsSecurityOption.BACKUP_WALLET)}
+                    selectedId={selectedSecurityOption}
+                  />
+                </div>
+              )}
```
Follow the same instruction above

ricardoak25 committed:
86a1gfwx5 - NEON3 - Implement Settings Screen Basic Structure (+4 squashed commits)
Squashed commits:
[97d6ca2a] 86a1gfwx5 - NEON3 - Implement Settings Screen Basic Structure
[0a3d7935] 86a1gfwx5 - NEON3 - Implement Settings Screen Basic Structure
[801ce690] 86a1gfwx5 - NEON3 - Implement Settings Screen Basic Structure
[24b581d1] 86a1gfwx5 - NEON3 - Implement Settings Screen Basic Structure
